sktime/utils/data_processing.py

```diff
@@ -335,36 +617,231 @@ def from_nested_to_3d_numpy(X):
 
     Parameters
     ----------
-    X : NumPy ndarray, input
+    X : np.ndarray
+        3-dimensional Numpy array to convert to nested Pandas DataFrame format
+
+    column_names: list-like, default = None
+        Optional list of names to use for naming nested DataFrame's columns
+
+    cells_as_numpy : bool, default = False
+        If True, then nested cells contain NumPy array
+        If False, then nested cells contain Pandas Series
+
 
     Returns
     -------
-    pandas DataFrame
+    df : pd.DataFrame
     """
     df = pd.DataFrame()
-    n_instances = X.shape[0]
-    n_variables = X.shape[1]
-    for variable in range(n_variables):
-        df["var_" + str(variable)] = [
-            pd.Series(X[instance][variable]) for instance in range(n_instances)
-        ]
+    # n_instances, n_variables, _ = X.shape
+    n_instances, n_columns, n_timepoints = X.shape
+
+    container = np.array if cells_as_numpy else pd.Series
+
+    if column_names is None:
+        column_names = _make_column_names(n_columns)
+
+    else:
+        if len(column_names) != n_columns:
+            msg = " ".join(
+                [
+                    f"Input 3d Numpy array as {n_columns} columns,",
+                    f"but only {len(column_names)} names supplied",
+                ]
+            )
+            raise ValueError(msg)
+
+    for j, column in enumerate(column_names):
+        df[column] = [container(X[instance, j, :]) for instance in range(n_instances)]
     return df
 
 
 def is_nested_dataframe(X):
-    return isinstance(X, pd.DataFrame) and isinstance(
-        X.iloc[0, 0], (np.ndarray, pd.Series)
-    )
+    """Checks whether the input is a nested DataFrame.
+
+    To allow for a mixture of nested and primitive columns types the
+    the considers whether any column is a nested np.ndarray or pd.Series.
+
+    By checking whether the first row has a nested structure, this implicitly
+    assumes that any column with nested structure will have that structure
+    in the first row. This will be true if column contains a homogenous
+    nested structure and can be true if it contains a mix of nested and
+    primitive types, but happens to have a nested structure in the first row.
+
+    Parameters
+    -----------
+    X :
+        Input that is checked to determine if it is a nested DataFrame.
+
+    Returns
+    -------
+    bool :
+        Whether the input is a nested DataFrame
+    """
+    # return isinstance(X, pd.DataFrame) and isinstance(
+    #     X.iloc[0, 0], (np.ndarray, pd.Series)
+    # )
+    is_dataframe = isinstance(X, pd.DataFrame)
+
+    # If not a DataFrame we know is_nested_dataframe is False
+    if not is_dataframe:
+        return is_dataframe
+
+    # Otherwise we'll see if any column has a nested structure in first row
+    else:
+        is_nested = are_columns_nested(X).any()
+
+        return is_dataframe and is_nested
+
+
+def nested_dataframes_equal(X1, X2):
```

Comment:
For this we already have a function [here](https://github.com/alan-turing-institute/sktime/blob/30b06140eb6e94b2845e6239efd679414dc13188/sktime/utils/_testing/estimator_checks.py#L641). Could you compare them and see if we can combine them into a single function if necessary?